File: packages/core/src/policies/draft.ts
/**
 * DefaultDraftPolicy — the T12 primitive.
 *
 * Decides whether a write on a published entity should redirect to drafts.
 * Also mints and verifies HMAC-signed preview tokens for rendering drafts
 * without authentication.
 *
 * Ported from Solid# `solid-backend/controllers/cms_pages.py:1193` (decision),
 * and the preview-token audit in `bounded-extraction-notes/02-t12-audit.md`.
 * Conforms to Bounded RFC v0 § 5.2 + § 9.3 (preview token MUSTs).
 *
 * Decision rule (evaluate):
 *
 *   1. entity is null (create op)                 -> allow
 *   2. entity does not implement IDraftableEntity -> allow  (policy inert)
 *   3. entity.is_published !== true               -> allow  (direct write OK)
 *   4. action.payload.publish === true            -> allow  (explicit live write)
 *   5. otherwise                                  -> redirect { target: 'draft', payload }
 *
 * Preview token format:
 *
 *   base64url( `v1.${entity_id}.${tenant_id}.${exp_ms}` + "." + hmac_sha256 )
 *
 * Verify MUST:
 *   - Use constant-time comparison on the HMAC
 *   - Reject expired / tampered / wrong-secret tokens
 *   - Enforce a max TTL of 168 hours (RFC v0 §9.3)
 */

import {
  createHmac,
  timingSafeEqual,
} from 'node:crypto';
import type {
  IDraftPolicy,
  IDraftableEntity,
  IPolicy,
  IPolicyContext,
  IPolicyDecision,
} from '../types.js';

const TOKEN_VERSION = 'v1';
const MAX_TTL_HOURS = 168; // RFC v0 §9.3 MUST
const DEFAULT_TTL_HOURS = 24;

export class DefaultDraftPolicy implements IDraftPolicy, IPolicy {
  readonly name = 'draft' as const;

  async evaluate(ctx: IPolicyContext): Promise<IPolicyDecision> {
    // 1. No entity (create) -> allow
    if (!ctx.entity) {
      return { kind: 'allow', policy: this.name };
    }

    // 2. Not a draftable entity -> policy inert
    if (!isDraftableEntity(ctx.entity)) {
      return { kind: 'allow', policy: this.name };
    }

    const entity = ctx.entity;

    // 3. Not published -> live writes are fine
    if (entity.is_published !== true) {
      return { kind: 'allow', policy: this.name };
    }

    // 4. Caller opted into a live write
    const payload = ctx.action.payload;
    if (isObject(payload) && payload['publish'] === true) {
      return { kind: 'allow', policy: this.name };
    }

    // 5. Redirect to draft
    return {
      kind: 'redirect',
      policy: this.name,
      target: 'draft',
      payload: ctx.action.payload ?? null,
    };
  }

  merge(entity: IDraftableEntity, draft_payload: unknown): IDraftableEntity {
    // Merge is a pure shape transformer: returns the entity with draft cleared
    // and with `draft_payload` represented as the new live state. The caller
    // supplies the real merge logic for their domain (we do not mutate columns
    // here because IDraftableEntity is abstract — actual field merge happens
    // in the domain layer / IDraftStore.publish).
    void draft_payload; // eslint-disable-line @typescript-eslint/no-unused-vars
    return {
      id: entity.id,
      tenant_id: entity.tenant_id,
      is_published: true,
      has_pending_draft: false,
      draft_updated_at: null,
      draft_updated_by: null,
    };
  }

  clear(entity: IDraftableEntity): IDraftableEntity {
    return {
      id: entity.id,
      tenant_id: entity.tenant_id,
      is_published: entity.is_published,
      has_pending_draft: false,
      draft_updated_at: null,
      draft_updated_by: null,
    };
  }

  signPreviewToken(
    entity_id: number | string,
    tenant_id: number | string,
    ttl_hours: number,
    secret: string,
  ): string {
    if (!secret) throw new Error('signPreviewToken: secret must be non-empty');
    const ttl = clampTtl(ttl_hours);
    const exp_ms = Date.now() + ttl * 3600_000;
    const payload = `${TOKEN_VERSION}.${entity_id}.${tenant_id}.${exp_ms}`;
    const mac = hmac(secret, payload);
    return toBase64Url(`${payload}.${mac}`);
  }

  verifyPreviewToken(
    token: string,
    secret: string,
  ): { entity_id: number | string; tenant_id: number | string } | null {
    if (!token || !secret) return null;
    let decoded: string;
    try {
      decoded = fromBase64Url(token);
    } catch {
      return null;
    }
    // Split from the right: HMAC may contain dots? No — hex only. Last segment is mac.
    const parts = decoded.split('.');
    if (parts.length !== 5) return null;
    const [version, entity_id_s, tenant_id_s, exp_ms_s, mac] = parts;
    if (version !== TOKEN_VERSION) return null;

    // Recompute HMAC and compare in constant time
    const expected_payload = `${version}.${entity_id_s}.${tenant_id_s}.${exp_ms_s}`;
    const expected_mac = hmac(secret, expected_payload);
    if (!constantTimeHexEquals(mac ?? '', expected_mac)) return null;

    // Expiry check
    const exp_ms = Number(exp_ms_s);
    if (!Number.isFinite(exp_ms) || exp_ms <= Date.now()) return null;

    // Narrow entity_id / tenant_id: preserve numeric shape when possible
    const entity_id = toIdMaybeNumeric(entity_id_s ?? '');
    const tenant_id = toIdMaybeNumeric(tenant_id_s ?? '');
    return { entity_id, tenant_id };
  }
}

// ----------------------------------------------------------------------------
// helpers — purposely exported as internal / not re-exported by index.ts
// ----------------------------------------------------------------------------

function isDraftableEntity(entity: object): entity is IDraftableEntity {
  return (
    'is_published' in entity &&
    typeof (entity as IDraftableEntity).is_published === 'boolean'
  );
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function clampTtl(hours: number): number {
  if (!Number.isFinite(hours) || hours <= 0) return DEFAULT_TTL_HOURS;
  if (hours > MAX_TTL_HOURS) return MAX_TTL_HOURS;
  return Math.floor(hours);
}

function hmac(secret: string, payload: string): string {
  return createHmac('sha256', secret).update(payload, 'utf8').digest('hex');
}

function constantTimeHexEquals(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  try {
    const ab = Buffer.from(a, 'hex');
    const bb = Buffer.from(b, 'hex');
    if (ab.length !== bb.length) return false;
    return timingSafeEqual(ab, bb);
  } catch {
    return false;
  }
}

function toBase64Url(s: string): string {
  return Buffer.from(s, 'utf8')
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

function fromBase64Url(s: string): string {
  const pad = (4 - (s.length % 4)) % 4;
  const b64 = s.replace(/-/g, '+').replace(/_/g, '/') + '='.repeat(pad);
  return Buffer.from(b64, 'base64').toString('utf8');
}

function toIdMaybeNumeric(s: string): number | string {
  if (/^-?\d+$/.test(s)) {
    const n = Number(s);
    if (Number.isFinite(n) && Number.isSafeInteger(n)) return n;
  }
  return s;
}

File: python/src/themis/policies/draft.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Adam Campbell
"""DefaultDraftPolicy — the T12 primitive (RFC v0 § 5.2, § 9.3).

Decision rule:
  1. entity is None (create)                → allow
  2. entity is not draftable                → allow  (inert)
  3. entity.is_published is not True        → allow
  4. action.payload['publish'] is True      → allow  (explicit live write)
  5. otherwise                              → redirect {target: 'draft', payload}

Preview token — byte-identical to ``@themis/core``:
  base64url( "v1.<entity_id>.<tenant_id>.<exp_ms>" + "." + hmac_sha256_hex )
Verify uses constant-time comparison, rejects expired/tampered/wrong-secret/
wrong-version tokens, and clamps TTL to 168 hours.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import math
import re
import time
from typing import Any, Mapping, Optional, Union

from ..types import Allow, Decision, DraftableEntity, Id, PolicyContext, Redirect

TOKEN_VERSION = "v1"
MAX_TTL_HOURS = 168
DEFAULT_TTL_HOURS = 24
_INT = re.compile(r"^-?\d+$")


def _b64url(s: str) -> str:
    return base64.urlsafe_b64encode(s.encode("utf-8")).decode("ascii").rstrip("=")


def _from_b64url(s: str) -> str:
    pad = (4 - len(s) % 4) % 4
    return base64.urlsafe_b64decode((s + "=" * pad).encode("ascii")).decode("utf-8")


def _hmac(secret: str, payload: str) -> str:
    return hmac.new(secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).hexdigest()


def _clamp_ttl(hours: float) -> int:
    if not isinstance(hours, (int, float)) or math.isnan(hours) or math.isinf(hours) or hours <= 0:
        return DEFAULT_TTL_HOURS
    if hours > MAX_TTL_HOURS:
        return MAX_TTL_HOURS
    return int(hours)


def _id_maybe_numeric(s: str) -> Id:
    if _INT.match(s):
        n = int(s)
        if -(2**53 - 1) <= n <= 2**53 - 1:
            return n
    return s


class DefaultDraftPolicy:
    name = "draft"

    def evaluate(self, ctx: PolicyContext) -> Decision:
        entity = ctx.entity
        if entity is None or not isinstance(entity, DraftableEntity):
            return Allow(policy=self.name)
        if entity.is_published is not True:
            return Allow(policy=self.name)
        payload = ctx.action.payload
        if isinstance(payload, Mapping) and payload.get("publish") is True:
            return Allow(policy=self.name)
        return Redirect(policy=self.name, target="draft", payload=payload if payload is not None else None)

    def merge(self, entity: DraftableEntity, draft_payload: Any) -> DraftableEntity:
        return DraftableEntity(
            id=entity.id, tenant_id=entity.tenant_id, is_published=True,
            has_pending_draft=False, draft_updated_at=None, draft_updated_by=None,
        )

    def clear(self, entity: DraftableEntity) -> DraftableEntity:
        return DraftableEntity(
            id=entity.id, tenant_id=entity.tenant_id, is_published=entity.is_published,
            has_pending_draft=False, draft_updated_at=None, draft_updated_by=None,
        )

    def sign_preview_token(self, entity_id: Id, tenant_id: Id, ttl_hours: float, secret: str) -> str:
        if not secret:
            raise ValueError("sign_preview_token: secret must be non-empty")
        exp_ms = int(time.time() * 1000) + _clamp_ttl(ttl_hours) * 3_600_000
        payload = f"{TOKEN_VERSION}.{entity_id}.{tenant_id}.{exp_ms}"
        return _b64url(f"{payload}.{_hmac(secret, payload)}")

    def verify_preview_token(self, token: str, secret: str) -> Optional[dict[str, Id]]:
        if not token or not secret:
            return None
        try:
            decoded = _from_b64url(token)
        except Exception:  # noqa: BLE001 — malformed token is a None, never a raise
            return None
        parts = decoded.split(".")
        if len(parts) != 5:
            return None
        version, entity_s, tenant_s, exp_s, mac = parts
        if version != TOKEN_VERSION:
            return None
        expected = _hmac(secret, f"{version}.{entity_s}.{tenant_s}.{exp_s}")
        if len(mac) != len(expected) or not hmac.compare_digest(mac, expected):
            return None
        try:
            exp_ms = int(exp_s)
        except ValueError:
            return None
        if exp_ms <= int(time.time() * 1000):
            return None
        return {"entity_id": _id_maybe_numeric(entity_s), "tenant_id": _id_maybe_numeric(tenant_s)}


__all__ = ["DefaultDraftPolicy"]
